tree/tree.go

```diff
@@ -0,0 +1,160 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return &Tree{
+		t:  tree,
+		db: database,
+	}, nil
+}
+
+func (t *Tree) DB() db.Database {
+	return t.db
+}
+
+// Get returns the value for a given key.
+func (t *Tree) Get(rTx db.ReadTx, key []byte) ([]byte, error) {
+	_, v, err := t.t.GetWithTx(rTx, key)
+	if err != nil {
+		return nil, err
+	}
+
+	return v, nil
+}
+
+// Set adds or updates a key. First performs a Get, and if the key does not
+// exist yet it will call Add, while if the key already exists, it will perform
+// an Update. If the non-existence or existence of the key is already known, is
+// more efficient to directly call Add or Update.
+func (t *Tree) Set(wTx db.WriteTx, key, value []byte) error {
+	_, err := t.Get(wTx, key)
+	if err == arbo.ErrKeyNotFound {
+		// key does not exist, use Add
+		return t.Add(wTx, key, value)
+	} else if err != nil {
+		return err
+	}
+	// key already exists, use Update
+	return t.t.UpdateWithTx(wTx, key, value)
+}
+
+// Add adds a new leaf, if the key already exists will return error
+func (t *Tree) Add(wTx db.WriteTx, key, value []byte) error {
+	return t.t.AddWithTx(wTx, key, value)
+}
+
+// AddBatch adds a batch of key-values to the Tree. Returns an array containing
+// the indexes of the keys failed to add. Supports empty values as input
+// parameters, which is equivalent to 0 valued byte array.
+func (t *Tree) AddBatch(wTx db.WriteTx, keys, values [][]byte) ([]int, error) {
+	return t.t.AddBatchWithTx(wTx, keys, values)
+}
+
+func (t *Tree) Iterate(callback func(key, value []byte) bool) error {
+	callbackWrapper := func(currLvl int, k, v []byte) bool {
+		callback(k, v)
+		return true
+	}
+	rTx := t.db.ReadTx()
+	root, err := t.Root(rTx)
+	if err != nil {
+		return err
+	}
+	return t.t.IterateWithStop(root, callbackWrapper)
+}
+
+// Root returns the current root.
+func (t *Tree) Root(rTx db.ReadTx) ([]byte, error) {
+	return t.t.RootWithTx(rTx)
+}
+
+// Size returns the number of leafs under the current root
+func (t *Tree) Size(rTx db.ReadTx) uint64 {
+	// TODO we need to review where this would be used and in which
+	// context, and define the way how this is computed
+	return 0
+}
+
+// GenProof returns a byte array with the necessary data to verify that the
+// key&value are in a leaf under the current root
+func (t *Tree) GenProof(rTx db.ReadTx, key []byte) ([]byte, error) {
+	// TODO NOTE leafK & leafV will be needed for verifying the proof,
+	// maybe is worth to return them
+	_, _, s, existence, err := t.t.GenProofWithTx(rTx, key)
+	if err != nil {
+		return nil, err
+	}
+	if !existence {
+		// proof of non-existence currently not needed in vocdoni-node
+		return nil, fmt.Errorf("key does not exist %s", hex.EncodeToString(key))
+	}
+	return s, nil
+}
+
+// VerifyProof checks the proof for the given key, value and root, using the
+// hash function of the Tree
+func (t *Tree) VerifyProof(key, value, proof, root []byte) (bool, error) {
```

Comment:
I think it would be nice if we find a way to allow calling this function without using an already opened tree.  For example, I suggest adding a method like this:
```
func VerifyProof(hashFunc HashFunction, key, value, proof, root []byte) (bool, error)
```
And `Tree.VerifyProof` just calls it.  This way we can verify proofs without an opened tree.

Reply:
Agree, thought about this but stuck to the spec. Added the function now